tensorflow/core/ops/resource_variable_ops.cc

```diff
@@ -30,15 +32,24 @@ namespace tensorflow {
 namespace {
 
 Status ReadVariableShapeFn(InferenceContext* c) {
-  std::vector<ShapeAndType> shape_and_type;
-  TF_RETURN_IF_ERROR(
-      shape_inference::ValidateVariableResourceHandle(c, &shape_and_type));
-  c->set_output(0, shape_and_type[0].shape);
-  if (shape_and_type[0].dtype == DT_VARIANT && shape_and_type.size() > 1) {
-    std::vector<ShapeAndType> variant_shape_and_type;
-    std::copy(shape_and_type.begin() + 1, shape_and_type.end(),
-              std::back_inserter(variant_shape_and_type));
-    c->set_output_handle_shapes_and_types(0, variant_shape_and_type);
+  // Return annotated shape if found.
```

Comment:
It would be nice to extend this comment with information about the new annotation. E.g.: The user can add a "_shape" attribute to ReadVariableOp nodes, and if that attribute is found, then shape inference will return the shape stored there. This is useful for inferring shapes in a function, where the user can annotate the graph using the variable capture list of the function.